### ISSUE TITLE
There is a input sequence, how to calculate probability for a special output sequence?

### ISSUE THREAD

**Issue by yick2232** [tensorflow/tensorflow#11602]

**Comment by yick2232** [comment]
Hi, i trained a model of seq2seq, the result sequence of decoder is so good, but I try to calculate the probability for a special outputs sequence, I found it's so hard for me.

**Comment by eth-n** [comment]
The first thing I would do is to try and define your probability space. If you impose no requirements on the input sequence, you would have to perform an exhaustive search on your probability space. That would look something like `aaaaa`, `aaaab`, `aaaac`... for a five character input sequence. If you're accounting for simple punctuation (maybe `,.?!'` and space), upper and lower case letters, and numbers, this is about 70^5 (~1.7 * 10^9) possible five character input strings. Hundreds of millions of these are likely to be garbage strings that you wouldn't expect your model to ever encounter, but may still produce your desired output sequence! While interesting, the metric you're trying to find may not be as useful as you expect it to be unless you find a reasonable way of cutting down the sheer size of the probability space (which grows rather quickly).

**Comment by asimshankar** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by yick2232** [comment]
OK, I got it.

**Comment by yick2232** [comment]
@eth-n  I try to use Trie to solve the problem,  limit the output sequence in fixed set and to got a best output sequence from the fixed outputs set, but the effect of generation is so terrible. Can you give me some suggestions, thanks a lot.

**Comment by eth-n** [comment]
Sorry @yick2232, I was intrigued by the abstract premise of your question and haven't tried anything along these lines before. I imagine someone's tried seq2seq visualization before; try looking very carefully through stack overflow, or look for blogs on the topic.

One of my coworkers wrote this post on visualizing RNNs, perhaps it'll help you in some regard: https://medium.com/datalogue/attention-in-keras-1892773a4f22

**Comment by yick2232** [comment]
@eth-n Thank you very much ! That is my issue,  I'll find the solution from other platform.  Thanks (^_^)

**Comment by eth-n** [comment]
@yick2232 ran into this by chance this morning, seems like you can use these techniques to look at `p(y|D, H), the distribution over network outputs y given data and a model` https://www.cs.cmu.edu/afs/cs/academic/class/15782-f06/slides/bayesian.pdf

**Comment by yick2232** [comment]
Thanks, I already finished the book that you gave me. I got a lot, and I find there is a discussion in [other issue of seq2seq](https://github.com/google/seq2seq/issues/81). They solve my problem.

**Comment by nikita-solvvy** [comment]
Can you elaborate on how you solved your problem? I seems to have the exact question.

**Comment by yick2232** [comment]
Okay, my problem is how to calculate the probability of a special output string.  I can got a logits of all vocabulary when the decoder run once.  And then, got the probability of  a character when the decoder run a step, the character in the specified output string.  In generally, you can take the specified character probability by normalize the logits, and take the log of the normalization result. Note that you should assign the current specified character as input of next decoder. Finally, just sum the probability of every specified character,  because of you have take the log of them.
